pinotdb/db.py

```diff
@@ -53,23 +53,44 @@ def g(self, *args, **kwargs):
     return g
 
 
-def get_description_from_row(row):
+def get_description_from_rows(rows):
     """
-    Return description from a single row.
+    Return description by scraping the rows
 
     We only return the nameand  type (inferred from the data).
     """
+    types = {}
+    unknown_columns = 0
+    for row in rows:
+        for name, value in row.items():
+            if name not in types:
+                types[name] = None
+                unknown_columns += 1
+            current_type = types[name]
+            if value is not None:
+                new_type = get_type(value)
+                assert new_type is not None, f'No known type for {value}'
+                if current_type is None:
+                    types[name] = new_type
+                    unknown_columns -= 1
+                elif new_type is not current_type:
+                    raise exceptions.DatabaseError(
+                            f'Differing column type found for column {name}:'
+                            f'{current_type} vs {new_type}')
+        if unknown_columns <= 0:
+            break
+
     return [
         (
             name,               # name
-            get_type(value),    # type_code
+            type,               # type_code
             None,               # [display_size]
             None,               # [internal_size]
             None,               # [precision]
             None,               # [scale]
             None,               # [null_ok]
         )
-        for name, value in row.items()
+        for name, type in types.items()
```

Comment:
This will not work in Python < 3.6, since the dictionary is not guaranteed to be ordered and the cursor description might not match the actual order of the columns in the data. Can you instantiate `types` to an `OrderedDict`?

Reply:
Thank you ! Great catch. I missed the fact that we are returning an array and not a dict !. Fixing.